### PR TITLE
Add extrapolation feature to interp_hybrid_to_pressure

### LINKED ISSUE #207: Vinth2p_ecmwf
For the new AMP diagnostic framework we would really like to have the functionality of
Vinth2p_ecmwf to extrapolate values  below the ground. Using an ECMWF formulation.
The corresponding NCL function is here: 
https://www.ncl.ucar.edu/Document/Functions/Built-in/vinth2p_ecmwf.shtml

### PR DESCRIPTION
**Pull request by hCraker**
Closes #207 and #132  This PR adds functionality to the pre-existing `interp_hybrid_to_pressure` function to recreate NCL's [vinth2p_ecmwf](https://www.ncl.ucar.edu/Document/Functions/Built-in/vinth2p_ecmwf.shtml).

Todo:
- [x] create function
- [x] create unittests
- [x] write documentation
- [x] add function to init.py

### CODE REVIEW

File: src/geocat/comp/interpolation.py [hunk 1]
```diff
@@ -134,17 +134,93 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
+    tstar = t_bot * (1 + alpha * (ps / p_sfc - 1))
+    hgt = phi_sfc * g_inv
+    t0 = tstar + 0.0065 * hgt
+
+    alph = xr.where((tstar <= 290.5) & (t0 > 290.5),
+                    R_d / phi_sfc * (290.5 - tstar), alpha)
+
+    alph = xr.where((tstar > 290.5) & (t0 > 290.5), 0, alph)
+    tstar = xr.where((tstar > 290.5) & (t0 > 290.5), 0.5 * (290.5 + tstar),
+                     tstar)
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:173]: Same as above

File: src/geocat/comp/interpolation.py [hunk 2]
```diff
@@ -134,17 +134,93 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
+    tstar = xr.where((tstar < 255), 0.5 * (tstar + 255), tstar)
+
+    alnp = alph * np.log(lev / ps)
+    return hgt - R_d * tstar * g_inv * np.log(
+        lev / ps) * (1 + 0.5 * alnp + 1 / 6 * alnp**2)
+
+
+def _vertical_remap_extrap(new_levels, lev_dim, data, output, pressure, ps,
+                           variable, t_bot, phi_sfc):
+    # TODO: check for appropriate input values
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:184]: Is this TODO going to be addressed in this PR?

File: src/geocat/comp/interpolation.py [hunk 3]
```diff
@@ -134,17 +134,93 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
+                     tstar)
+
+    tstar = xr.where((tstar < 255), 0.5 * (tstar + 255), tstar)
+
+    alnp = alph * np.log(lev / ps)
+    return hgt - R_d * tstar * g_inv * np.log(
+        lev / ps) * (1 + 0.5 * alnp + 1 / 6 * alnp**2)
+
+
+def _vertical_remap_extrap(new_levels, lev_dim, data, output, pressure, ps,
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:182]: This function could also use a docstring

File: src/geocat/comp/interpolation.py [hunk 4]
```diff
@@ -134,17 +134,93 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
                               lev_dim: str = None,
-                              method: str = 'linear') -> xr.DataArray:
-    """Interpolate data from hybrid-sigma levels to isobaric levels. Keeps
-    attributes (i.e. meta information) of the input data in the output as
-    default.
+                              method: str = 'linear',
+                              extrapolate: bool = False,
+                              variable: str = None,
+                              t_bot: xr.DataArray = None,
+                              phi_sfc: xr.DataArray = None) -> xr.DataArray:
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:220]: Should we also include explicit support for anything array-like or are we only officially supporting numpy here?

File: src/geocat/comp/interpolation.py [hunk 5]
```diff
@@ -134,17 +134,93 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
+    """Interpolate and extrapolate data from hybrid-sigma levels to isobaric
+    levels. Keeps attributes (i.e. meta information) of the input data in the
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:222]: This is picky, feel free to disagree
```suggestion
    levels. Keeps attributes (i.e. metadata) of the input data in the
```

**Review by anissa111 — COMMENTED** [review]
Mostly surface level comments here. 

You'll also need to add your new functions to the user and internal api pages

File: src/geocat/comp/interpolation.py (hunk 2 above)
Comment (hCraker) [reply src/geocat/comp/interpolation.py:184]: I added some argument checks in the main interp_hybrid_to_pressure function, so this TODO isn't needed anymore. I'll remove it

File: src/geocat/comp/interpolation.py (hunk 4 above)
Comment (hCraker) [reply src/geocat/comp/interpolation.py:262]: I only wrote it to support xarray objects. I can add a conditional to make an `xr.DataArray` out of any array like input

File: src/geocat/comp/interpolation.py [hunk 6]
```diff
@@ -133,16 +133,177 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
+
+    p_sfc: :class:`xarray.DataArray`
+        The pressure at the lowest level of the model. Must be in the same units as ``lev`` and ``ps``
+
+    ps: :class:`xarray.DataArray`
+        An array of surface pressures. Must be in the same units as ``lev`` and ``p_sfc``
+
+    phi_sfc: :class:`xarray.DataArray`
+        The geopotential at the lowest level of the model.
+    """
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:161]: I think we still need to include the return here. Also, let me take a swing at rendering the equations with latex to make them more accessible

File: src/geocat/comp/interpolation.py [hunk 7]
```diff
@@ -133,16 +133,177 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
+
+    p_sfc: :class:`xarray.DataArray`
+        The pressure at the lowest level of the model. Must be in the same units as ``lev`` and ``ps``
+
+    ps : :class:`xarray.DataArray`
+        An array of surface pressures. Must be in the same units as ``lev`` and ``p_sfc``
+
+    phi_sfc:
+        The geopotential at the lowest level of the model.
+    """
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:206]: Same wrt returns section

File: src/geocat/comp/interpolation.py [hunk 8]
```diff
@@ -133,16 +133,177 @@ def _vertical_remap(func_interpolate, new_levels, xcoords, data, interp_axis=0):
...
+        "other", the value of ``data`` at the lowest model level will be used
+        as the below ground fill value. Required if extrapolate is True.
+
+    t_bot: :class:`xarray.DataArray`
+        Temperature at the lowest (bottom) level of the model.
+
+    phi_sfc:
+        The geopotential at the lowest level of the model.
+    Returns
+    -------
```
Comment (anissa111) [inline src/geocat/comp/interpolation.py:267]: Did you mean to leave the returns section blank here?

File: test/test_interpolation.py [hunk 9]
```diff
@@ -9,12 +9,12 @@
...
-    from src.geocat.comp import interp_multidim, interp_hybrid_to_pressure, \
-        interp_sigma_to_hybrid
-else:
-    from geocat.comp import interp_multidim, interp_hybrid_to_pressure, \
+#if "--cov" in str(sys.argv):
+from src.geocat.comp import interp_multidim, interp_hybrid_to_pressure, \
         interp_sigma_to_hybrid
+#else:
+#    from geocat.comp import interp_multidim, interp_hybrid_to_pressure, \
+#        interp_sigma_to_hybrid
```
Comment (anissa111) [inline test/test_interpolation.py:17]: Can we remove the comments here?

**Review by anissa111 — COMMENTED** [review]
Just a few questions about the returns sections. Also, I'm going to take a swing real quick at making the equations with latex. If it doesn't work though, absolutely let's just go ahead without that.

File: test/test_interpolation.py (hunk 9 above)
Comment (hCraker) [reply test/test_interpolation.py:17]: Oh whoops I totally forgot about that. That is my work around to get tests to run from the local files. Sorry I usually have more attention to detail

**Review by anissa111 — APPROVED** [review]
Looks good!! Let's get this merged and released :)

**Comment by katiedagon** [comment]
Hi @anissa111 @hCraker the added extrapolation feature looks super useful! Do you know if this is in a released version of geocat-comp yet? I tried to use it today (with geocat-comp version 2022.11.0) and I got an error:
`TypeError: interp_hybrid_to_pressure() got an unexpected keyword argument 'extrapolate'`

**Comment by anissa111** [comment]
> Hi @anissa111 @hCraker the added extrapolation feature looks super useful! Do you know if this is in a released version of geocat-comp yet? I tried to use it today (with geocat-comp version 2022.11.0) and I got an error:
> `TypeError: interp_hybrid_to_pressure() got an unexpected keyword argument 'extrapolate'`

Katie, it just missed the 2022.11.0 release, it will be in 2022.12.0!

**Comment by katiedagon** [comment]
Excellent, thanks! Do you have an idea of the timeline for that release?

**Comment by anissa111** [comment]
I'm not totally sure, but probably around mid month based on recent releases

**Comment by katiedagon** [comment]
@NCAR/geocat Great to see this feature released - thank you! @TeaganKing has a follow-up question about the requirements for using extrapolation, so I wanted to also flag that here.
https://github.com/NCAR/geocat-comp/issues/26#issuecomment-1423269880

**Pull request merged**